nmmo/task/task_api.py

```diff
@@ -1,81 +1,146 @@
+
+  @property
+  def assignee(self):
+    return self._assignee
+
+  @abstractmethod
+  def rewards(self, gs: GameState) -> Tuple[Dict[int, float], Dict[int, Dict]]:
+    """ Returns a mapping from ent_id to rewards and infos for all 
+    entities in assignee
+    """
+    raise NotImplementedError
+
+class PredicateTask(Task, ABC):
+  def __init__(self,
+               assignee: Group,
+               predicate: Predicate):
+    super().__init__(assignee)
+    self._predicate = predicate
+
+  def evaluate(self, gs: GameState) -> float:
+    name = self._predicate.name
+    cache = gs.cache_result
+    if name not in cache:
+      cache[name] = self._predicate(gs)
+    return cache[name]
+
+class Once(PredicateTask):
+  def __init__(self,
+               assignee: Group,
+               predicate: Predicate,
+               reward = 1):
+    super().__init__(assignee, predicate)
+    self._reward = reward
+    self._completed = False
+
+  def rewards(self, gs: GameState):
+    rewards = {int(ent_id): 0 for ent_id in self._assignee}
+    infos = {int(ent_id): {self._predicate.name: self.evaluate(gs)}
+             for ent_id in self._assignee}
+    if not self._completed and self.evaluate(gs):
+      self._completed = True
+      rewards = {int(ent_id): self._reward for ent_id in self._assignee}
+    return rewards, infos
+
+class Repeat(PredicateTask):
+  def __init__(self,
+               assignee: Group,
+               predicate: Predicate,
+               reward = 1):
+    super().__init__(assignee, predicate)
+    self._reward = reward
+
+  def rewards(self, gs: GameState):
+    rewards = {int(ent_id): 0 for ent_id in self._assignee}
+    infos = {int(ent_id): {self._predicate.name: self.evaluate(gs)}
+             for ent_id in self._assignee}
+    if self.evaluate(gs):
+      rewards = {int(ent_id): self._reward for ent_id in self._assignee}
+    return rewards, infos
 
+class MultiTask(Task):
+  def __init__(self, *tasks: Task):
+    assert len(tasks) > 0
+    super().__init__(reduce(lambda a,b: a.union(b),
+                            [task.assignee for task in tasks]))
+    self._tasks = tasks
+
+  def rewards(self, gs: GameState) -> Dict[int, float]:
+    rewards = {}
+    infos = {}
+    for task in self._tasks:
+      task_reward, task_infos = task.rewards(gs)
+      for ent_id, reward in task_reward.items():
+        rewards[ent_id] = rewards.get(ent_id,0) + reward
+      for ent_id, info in task_infos.items():
+        if not ent_id in infos:
+          infos[ent_id] = {}
+        infos[ent_id] = {**infos[ent_id], **info}
+
+    return rewards, infos
+
+################################################
+# Environment Wrapper
+# Eventually should replace env.py once stable
+# TODO(mark) Syllabus
 
-# TODO(kywch): how to divide the task system, into core vs. wrapper, etc.?
-#  We will address this once the task wrapper is stable
 # pylint: disable=abstract-method
-class TaskWrapper(nmmo.Env):
+class TaskEnv(nmmo.Env):
```

Comment:
If you want to use Syllabus you can make this subclass TaskWrapper instead. Alternatively, maybe I can just make a TaskEnv mixin for Syllabus That basically does what you have here.